gen/dcos-config.yaml

```diff
@@ -473,8 +473,6 @@ package:
       {
         "PATH": "/usr/bin:/bin",
         "SHELL": "/usr/bin/bash",
-        "LD_LIBRARY_PATH": "/opt/mesosphere/lib",
-        "SASL_PATH": "/opt/mesosphere/lib/sasl2",
```

Comment:
It appears that our bundle is lacking the SASL plugins - this means that IFF Mesos was trying to do SASL based authentication (e.g. CRAM-MD5), it would fail as it was unable to locate the needed plugin within the bundle and due to this line also on the DCOS system itself.
However, we commonly do not use SASL at all within DCOS, hence this may be entirely fine. We just need to be aware.

Reply:
This change I inherited from #707 without context – from what I am able to gather, though, it seems like we it will break two reasons:

1. [SASL is the default authentication strategy for Mesos](http://mesos.apache.org/documentation/latest/authentication/)
2. Unlike the linker's support for library-scoped pathing (i.e. `rpath` and friends), Cyrus SASL appears to look only in the single directory identified by [`SASL_PATH`](https://www.cyrusimap.org/sasl/sasl/reference/manpages/library/sasl_client_init.html?highlight=sasl_path), falling back to `/usr/lib/sasl2` if that's not set.

The alternative would be to keep this change and add a build phase to DC/OS that packages libsasl2 directly, as it looks like the configure script has `--with-plugindir` and `--with-configdir` that would allow a DC/OS-packaged libsasl2 to do the right thing without setting the environment variable.

I'm looking into building a DC/OS vendored sasl2, but I wouldn't mind a pointer or two if they're handy.

Reply:
re 1: Yes it is.
re 2: Correct, these mechanism plugins are dynamically loaded by libsasl itself and hence are not a hard dependency. They are only used when the specific mechanisms (e.g. CRAM-MD5) are used/requested

Let's take a step back here :) .... (ow how I commonly dislike such statements) ....

SASL is being used by Mesos for two things, framework- and agent-authentication. In the open-source world, currently only CRAM-MD5 is being used as a mechanism.

The Mesos agent is a core part of DC/OS and hence it has to be relying on the components installed within DC/OS - think of Mesos like the DC/OS kernel. `libsasl` while being a dependency of Mesos itself however is not considered a core component but only a dependency.

Frameworks on the other hand are not a core part of DC/OS, they are what I would describe as user-land software. Frameworks should not rely on any dependency other than the kernel (libmesos) within DC/OS. 

If a Framework now wants to use SASL for authenticating against the master, this framework should bring its own set of SASL plugins. While this may sound a bit picky or even non-sensical, this is what we commonly strife for -- one core argument here is that we don't want our user-land software to make assumptions about non kernel components as that would restrict us a lot in upgrading kernel dependencies. 

In the end, SASL as being a dependency of Mesos itself needs to work for both agents and masters without further user intervention. As such, we need `libsasl` to be able to locate its plugins and hence removing `SASL_PATH` is not a good idea unless of cause we build `libsasl` the way you are suggesting (`--with-plugindir`).

This still means that framework authors should not rely on any functioning SASL components on the DC/OS platform itself -- they should bundle their own libsasl + plugins.

In the end, this means I am suggesting to:
- leave `SASL_PATH` in or compile SASL the way you are suggesting
- tell our users that they should bundle `libsasl` and at least the CRAM-MD5 SASL mechanism plugin as part of their Mesos frameworks, if they intend to rely on this authentication variant

Reply:
Regarding "framework authors should not rely on any functioning SASL ..." -- does it make sense to bundle SASL libraries in the libmesos-bundle and have them use it? Would it be reasonable?

Reply:
many thanks @tillt for the additional information! As you suggest, I've backed out the SASL_PATH change for now; if I discover a way to vendor in libsasl in my copious free time I'll be sure to open a new PR.

Reply:
@karya0 it would make sense in my opinion, yes. There will likely be very very few users of that functionality but given that Mesos itself offers it and DC/OS does not prevent it, we should enable it via the libmesos-bundle.